File: src/main/java/net/alexwells/roomery/util/WorldUtils.java
package net.alexwells.roomery.util;

import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockAccess;

public class WorldUtils {
    public static <T extends TileEntity> T getTileEntity(IBlockAccess world, BlockPos pos) {
        TileEntity tileEntity = world.getTileEntity(pos);

        try {
            return (T) tileEntity;
        } catch (ClassCastException ignored) {

        }

        return null;
    }
}

File: src/main/java/net/alexwells/roomery/util/WorldUtils.kt
package net.alexwells.roomery.util

import net.minecraft.tileentity.TileEntity
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IBlockReader
import org.apache.logging.log4j.LogManager

fun <T : TileEntity> IBlockReader.getTile(pos: BlockPos): T? {
    val tileEntity = this.getTileEntity(pos)

    try {
        return tileEntity as T?
    } catch (ex: ClassCastException) {
        LogManager.getLogger().error("Trying to cast world's tile entity into needed type.")
        LogManager.getLogger().catching(ex)
    }

    return null
}
